gfx.h

```diff
@@ -5900,6 +5862,74 @@ class GfxInternal
+                    void *shader_identifier = state_object_properties->GetShaderIdentifier(sbt_record.shader_identifier_.c_str());
+                    memcpy((uint8_t *)upload_buffer.data_ + upload_buffer_offset, shader_identifier, D3D12_SHADER_IDENTIFIER_SIZE_IN_BYTES);
+                    upload_buffer_offset += D3D12_SHADER_IDENTIFIER_SIZE_IN_BYTES;
+                    auto local_root_signature_association = kernel.local_root_signature_associations_.find(sbt_record.shader_identifier_);
+                    if(local_root_signature_association != kernel.local_root_signature_associations_.end())
+                    {
+                        for(auto &parameter : sbt_record.bound_parameters_)
+                        {
+                            if(parameter.type_ >= Kernel::Parameter::kType_Count) continue;
+                            switch(parameter.type_)
+                            {
+                            case Kernel::Parameter::kType_Constants:
+                            case Kernel::Parameter::kType_ConstantBuffer:
+                                populateRootConstants(program, *sbt_record.parameters_, parameter,
+                                    (uint32_t *)((uint8_t *)upload_buffer.data_ + upload_buffer_offset), true);
+                                upload_buffer_offset += parameter.variable_size_;
+                                break;
+                            default:
+                                uint32_t descriptor_slot = (parameter.descriptor_slot_ != 0xFFFFFFFFu
+                                                                ? parameter.descriptor_slot_
+                                                                : dummy_descriptors_[parameter.type_]);
+                                if (parameter.descriptor_slot_ != 0xFFFFFFFFu)
+                                    initDescriptorParameter(program, invalidate_sbt_descriptors, parameter, descriptor_slot);
+                                for(uint32_t j = 0; j < parameter.descriptor_count_; ++j)
+                                {
+                                    auto descriptor_handle =
+                                        descriptors_.getGPUHandle(descriptor_slot + j);
+                                    memcpy((uint8_t *)upload_buffer.data_ + upload_buffer_offset,
+                                        &descriptor_handle, sizeof(descriptor_handle));
+                                    upload_buffer_offset += sizeof(descriptor_handle);
+                                }
+                                break;
+                            }
+                        }
+                    }
+                    command_list_->CopyBufferRegion(sbt_buffer.resource_, dst_offset, upload_buffer.resource_,
+                        src_offset, upload_buffer_offset - src_offset);
+                    sbt_record.commited_id_ = sbt_record.id_;
+                }
+                destroyBuffer(upload_gfx_buffer);
+            }
+        }
```

Comment:
This seems to be a copy/paste of the block above.

It probably would be best to factor this, what do you think?

Reply:
Indeed, the second block looks pretty much similar to previous one, but we can't merge them into just one loop, as the first loop intended to allocate descriptors, and if during the allocation we ran out of free descriptors, we allocate a new descriptor heap and populate it in the second loop, actually it inherits logic from regular descriptor binding population, just for SBT.